prov/util/src/util_attr.c

```diff
@@ -1017,13 +1026,18 @@ static void fi_alter_domain_attr(struct fi_domain_attr *attr,
 				 const struct fi_domain_attr *hints,
 				 uint64_t info_caps, uint32_t api_version)
 {
-	ofi_mr_mode_adjust(info_caps, &attr->mr_mode);
-
-	if (FI_VERSION_LT(api_version, FI_VERSION(1, 5)))
-		attr->mr_mode = attr->mr_mode & OFI_MR_BASIC_MAP ?
-			FI_MR_BASIC : FI_MR_SCALABLE;
-	else
-		attr->mr_mode &= ~(FI_MR_BASIC | FI_MR_SCALABLE);
+	int hints_mr_mode;
+
+	hints_mr_mode = hints ? hints->mr_mode : 0;
+	if (hints_mr_mode & (FI_MR_BASIC | FI_MR_SCALABLE)) {
+		attr->mr_mode = hints_mr_mode;
+	} else if (FI_VERSION_LT(api_version, FI_VERSION(1, 5))) {
+		attr->mr_mode = (attr->mr_mode && attr->mr_mode != FI_MR_SCALABLE) ?
+				FI_MR_BASIC : FI_MR_SCALABLE;
+	} else {
+		attr->mr_mode = ofi_mr_mode_adjust(info_caps,
+					attr->mr_mode & hints_mr_mode);
```

Comment:
We need to adjust the needed mr_mode bits based on the app's requested capabilities.  This removes unnecessary mr_mode bits based on the request.  If we want to remove this call, then we also need to remove it from the ofi_check_mr_mode() call.

Reply:
See #3442. If a provider only supports basic memory registration, it only supports virtual addressing. Although capabilities might indicate that certain bits may be unnecessary, does not change the functional requirements of the provider.

Reply:
I get that.  The alternative is to reject the fi_getinfo request completely even though the provider is capable of supporting it based on the features being requested.  Virtual addressing is irrelevant if we're not the target of an RMA/atomic operation.

Reply:
> The alternative is to reject the fi_getinfo request completely even though the provider is capable of supporting it based on the features being requested.

Why should the request be rejected? The request can be sufficiently satisfied. I believe that a relationship is being created here that is not valid. 

> Virtual addressing is irrelevant if we're not the target of an RMA/atomic operation.

I believe that memory registration requirements and requested libfabric data transfer capabilities are orthogonal topics, and here is why: Memory registration describes how my memory and remote memory is addressed. It affects whether my provider is responsible for giving a key to the application and whether key exchange must take place. It affects CQ event generation and other facets. However, requested libfabric capabilities only describe to me what an application intends to do with my fabric. Even if the local process will never be the target of a RMA or atomic, remote processes may be targets of my local process's RMA or atomic operations. 

Let's even get away from virtual addressing for a moment and talk about the fact that the FI_MR_PROV_KEY is also being stripped. It doesn't matter that the local rank isn't the target of an RMA or atomic, if the local rank would still initiate a transfer TO a different rank. The remote rank is going to have FI_REMOTE_READ or FI_REMOTE_WRITE... and would have basic registration(based on the ongoing assumptions on info contents), and expect key exchange. However, the local rank has different information about memory registration because those bits were stripped by the utility code. The application would expect no key exchange, because it can select well known keys (!FI_MR_PROV_KEY). The application would expect to use offset based addressing against a key it believe to have been registered using a fabric-registration method for scalable addressing (also not true if the remote rank believe it is using basic registration).

If we continue along this train of thought and apply it to verbs (abstractly), the rank that is capable of being the target of an RMA and/or atomic returns an fi_info structure that should contain the equivalent of FI_MR_BASIC. The other rank which is not going to be the target of an RMA and/or atomic returns an fi_info structure that should contain the equivalent of FI_MR_SCALABLE or 0. The first problem... is that the verbs provider already knows it can't do scalable and would have returned FI_ENODATA if it had done a final check against the altered data. The second problem is that once the rank with the FI_MR_SCALABLE fi_info structure attempts to open the fabric, it will or should observe -FI_EINVAL or something functionally equivalent because verbs only supports basic registration. This is a non-uniform result across nodes due to a conversion performed by utility code.